sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/JavaTypeInference.scala

```diff
@@ -131,28 +132,38 @@ object JavaTypeInference {
               s"of class $other")
         }
 
-        // TODO: we should only collect properties that have getter and setter. However, some tests
-        // pass in scala case class as java bean class which doesn't have getter and setter.
-        val properties = getJavaBeanReadableProperties(other)
-        val fields = properties.map { property =>
-          val returnType = typeToken.method(property.getReadMethod).getReturnType
-          val (dataType, nullable) = inferDataType(returnType, seenTypeSet + other)
-          new StructField(property.getName, dataType, nullable)
+        val fields = getObjectProperties(other).map {
+          case (propertyName, getterMethod, setterMethod) =>
+            val (dataType, nullable) = inferDataType(
+              TypeToken.of(getterMethod.getGenericReturnType),
+              seenTypeSet + other)
+            new StructField(propertyName, dataType, nullable)
         }
         (new StructType(fields), true)
     }
   }
 
-  def getJavaBeanReadableProperties(beanClass: Class[_]): Array[PropertyDescriptor] = {
-    val beanInfo = Introspector.getBeanInfo(beanClass)
-    beanInfo.getPropertyDescriptors.filterNot(_.getName == "class")
-      .filterNot(_.getName == "declaringClass")
-      .filter(_.getReadMethod != null)
-  }
-
-  private def getJavaBeanReadableAndWritableProperties(
-      beanClass: Class[_]): Array[PropertyDescriptor] = {
-    getJavaBeanReadableProperties(beanClass).filter(_.getWriteMethod != null)
+  /**
+   * Returns: Array[(porpertyName, getterName, setterName, propertyType)]
+   *
+   * Properties of the object are defined by a getter 'propertyType [get]PropertyName()' and a
+   * setter 'void [set]PropertyName(propertyType value)' functions; where [get]PropertyName is
+   * the name of the getter function, and [set]PropertyName is the name of the setter function.
+   */
+  def getObjectProperties(beanClass: Class[_]): Array[(String, Method, Method)] = {
+    for {
+      a <- beanClass.getMethods.filter(method => method.getParameterCount == 0)
+      b <- beanClass.getMethods.filter(method => method.getReturnType == Void.TYPE &&
+        method.getParameterCount == 1)
+      if (a.getName == b.getName ||
+        (a.getName.indexOf("get") == 0 && b.getName.indexOf("set") == 0 &&
+          a.getName.substring(3) == b.getName.substring(3))) &&
```

Comment:
It is a bit hacky here, compared with https://github.com/apache/spark/pull/24299

Reply:
Yes, I will try to simplify

We need to check that the getter and setter functions are either prefixed with get and set respectively or both functions are not prefixed

I did this modification in https://github.com/apache/spark/pull/24299

I will try to make the refactor the code if possible